verl/workers/rollout/sglang_rollout/sglang_rollout.py

```diff
@@ -1427,10 +1498,25 @@ async def generate(
         image_data: Optional[list[Any]] = None,
     ) -> torch.Tensor:
         """Generate sequence with token-in-token-out."""
-        request_sampling_params = self.sampling_params.copy()
-        request_sampling_params.update(sampling_params)
-        output = await self._handle_engine_generate(prompt_ids, request_sampling_params, image_data=image_data)
-        return output["output_ids"]
+        with ExitStack() as stack:
+            request_sampling_params = self.sampling_params.copy()
+            request_sampling_params.update(sampling_params)
+            awaitable = self._handle_engine_generate(
+                prompt_ids, request_sampling_params, image_data=image_data, req_id=request_id
+            )
+            task = asyncio.ensure_future(awaitable)
+            self.active_req[request_id] = task
+            stack.callback(lambda: self.active_req.pop(request_id, None))
+            output = await task
+            return output["output_ids"]
+
+    async def cancel_req(self, request_id: str):
+        logger.debug(f"cancel request {request_id} from cancel_req")
+        if request_id in self.active_req:
+            self._engine.tokenizer_manager.abort_request(request_id)
```

Comment:
I think it's safe to abort request which is already done, hence we can eliminate `self.active_req`.